userguide/analytics/userguide_analytics.adoc

```diff
@@ -1,63 +1,595 @@
+
+== Plugin Installation
+
+A plugin can be installed via the https://github.com/killbill/killbill-cloud/blob/master/kpm[Kill Bill Package Manager (KPM)]. For this, please follow the steps given below:
+
+. Ensure that you have kpm installed as explained https://github.com/killbill/killbill-cloud/tree/master/kpm#kpm-installation[here].
+
+. Install the plugin using the following `kpm` command (Replace `<path_to_install_plugin>` with the path where you want to install the plugin. This path should match the path specified by the `org.killbill.osgi.bundle.install.dir` property in the  https://docs.killbill.io/latest/userguide_configuration.html#_configuration_properties_table[Kill Bill Configuration]):
+[source,bash]
+kpm install_java_plugin analytics --destination=<path_to_install_plugin>
+
+. Verify that the plugin is installed properly. This can be done by ensuring that the plugin JAR is present at the `path_to_install_plugin` specified in the previous step. Alternatively, you can also open https://docs.killbill.io/latest/userguide_kaui.html[Kaui], hover over the plug icon and click on `analytics`. This should display the following screen:
+image:https://github.com/killbill/killbill-docs/raw/v3/userguide/assets/img/analytics-plugin/analytics-plugin-in-kaui.png[align=center]
+
+. If you would like to make changes to the plugin, you can clone the plugin code, build and deploy it as explained https://github.com/killbill/killbill-analytics-plugin[here].
+
+
+== Database Configuration
+
+As mentioned earlier, the analytics plugin uses its own reporting tables. To create these tables, please follow the steps given below:
+
+. Connect to the Kill Bill database.
+
+. Run the analytics plugin https://github.com/killbill/killbill-analytics-plugin/blob/0a25c18c3951b663e4d1ebf81619f325175209cf/src/main/resources/org/killbill/billing/plugin/analytics/ddl.sql[DDL].
+
+The <<analytics_tables, analytics tables>> section provides an overview of some of the important tables in the analytics database.
+
+== Plugin Configuration
+
+You can configure the following properties related to the analytics plugin in the https://docs.killbill.io/latest/userguide_configuration.html#global_configuration_properties[Kill Bill Configuration File]:
+
+[options="header",cols="1,1,1"]
+|===
+|Property Name   |Description   |Default Value   
+//----------------------
+|org.killbill.notificationq.analytics.tableName   |Specifies the name of the notification queue table   |notifications (Same as Kill Bill notification queue table)
+|org.killbill.notificationq.analytics.historyTableName   |Specifies the name of the notification queue history table    |notifications_history (Same as Kill Bill notification queue history table)
+|org.killbill.analytics.lockSleepMilliSeconds   |Specifies how long to sleep between retries to get the lock   |100   
```

Comment:
I see "the lock" here and in several other places. I don't know what it refers to, probably because of my lack of technical knowledge. Is there a need to be more specific about what the lock applies to?

Reply:
TBH, I'm not completely sure how the lock works so I've just added these descriptions based on the code comments which is why they may seem inadequate. That being said, I think configuring the lock is an advanced option, in most cases, the default values provided by the analytics plugin should be good enough. @pierre - any thoughts?

Reply:
Some critical operations in Kill Bill use a global lock (using the database) to ensure a single Kill Bill node performs a given operation at a time (the code is here: https://github.com/killbill/killbill-commons/tree/master/locker).

Bumping the `lockSleepMilliSeconds` parameter could help if these critical operations (refresh) take a long time: https://github.com/killbill/killbill-analytics-plugin/blob/095794e283c9466b89c397ea7a77c845875c30e7/src/main/java/org/killbill/billing/plugin/analytics/AnalyticsActivator.java#L114

Reply:
Thanks for clarifying @pierre. The `lockAttemptRetries` and `rescheduleIntervalOnLockSeconds` (configured on a per-tenant basis) also seem to be related to the lock. Under what circumstances would a user need to configure these to values different from the defaults provided by the plugin?

Reply:
By default, if a refresh is in progress (i.e. lock is taken) and another node tries to refresh the same account, it will block
attempt `lockAttemptRetries` times to grab the lock, sleeping `lockSleepMilliSeconds` in between retries, before giving up.

If `rescheduleIntervalOnLockSeconds` is configured, if the lock is taken, it will reschedule itself `rescheduleIntervalOnLockSeconds` in the future, before trying the above algorithm.

See https://github.com/killbill/killbill-analytics-plugin/blob/59c282f97960dc59b944ca33e17353bc19d7c842/src/main/java/org/killbill/billing/plugin/analytics/AnalyticsListener.java#L314-L339

Tweaking these values depend in practice on how often refreshes are scheduled, the size of accounts, etc.

Reply:
Thanks, I've added a small paragraph related to the lock based on the above.